allennlp/nn/beam_search.py

```diff
@@ -213,15 +214,33 @@ def _search(
         # Else, select the top `beam_size` tokens.
         # shape: (batch_size, beam_size), (batch_size, beam_size)
         if self.sampler is not None:
-            start_top_log_probabilities, start_predicted_classes = self.sampler(
-                start_class_log_probabilities, num_samples=self.beam_size
-            )
+            if isinstance(self.sampler, GumbelMaxSampler):
+                # phi represents the summed log probabilities of the sequence at hand
+                phi = start_class_log_probabilities
+
+                # Our sampler gets the perturbed log probabilities for each token
+                start_top_log_probabilities, start_predicted_classes = self.sampler(
+                    phi, num_samples=self.beam_size
+                )
+            else:
+                start_top_log_probabilities, start_predicted_classes = self.sampler(
+                    start_class_log_probabilities, num_samples=self.beam_size
+                )
```

Comment:
Is this special case handling for `GumbelMaxSampler` still necessary?

And it looks like the way we call `self.sampler()` in the `else` branch is incorrect / inconsistent with the signature of `Sampler`.